nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/MonitorActivity.java

```diff
@@ -217,224 +230,221 @@ protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+            localFlowActivityInfo.update(commonFlowActivityInfo);
+
+            try {
+                commonFlowActivityInfo.update(localFlowActivityInfo);
+                localFlowActivityInfo.setNextSyncMillis();
+            } catch (final SaveSharedFlowStateException ex) {
+                logger.debug("Failed to update common state.", ex);
             }
+        }
+    }
 
-            if (sendInactiveMarker && shouldThisNodeReport(isClusterScope, shouldReportOnlyOnPrimary, context)) {
-                lastInactiveMessage.set(System.currentTimeMillis());
+    private void onTriggerInactiveFlow(ProcessContext context, ProcessSession session, boolean isClusterScope, long lastActivity) {
+        final ComponentLog logger = getLogger();
+        final boolean shouldReportOnlyOnPrimary = shouldReportOnlyOnPrimary(isClusterScope, context);
+        final boolean shouldThisNodeReport = shouldThisNodeReport(isClusterScope, shouldReportOnlyOnPrimary, context);
 
-                FlowFile inactiveFlowFile = session.create();
-                inactiveFlowFile = session.putAttribute(inactiveFlowFile, "inactivityStartMillis", String.valueOf(previousSuccessMillis));
-                inactiveFlowFile = session.putAttribute(inactiveFlowFile, "inactivityDurationMillis", String.valueOf(now - previousSuccessMillis));
+        if (shouldThisNodeReport) {
+            sendInactivityMarker(context, session, lastActivity, logger);
+        }
+        lastInactiveMessage.set(System.currentTimeMillis());
+        setInactivityFlag(context.getStateManager());
+    }
 
-                final byte[] outBytes = context.getProperty(INACTIVITY_MESSAGE).evaluateAttributeExpressions(inactiveFlowFile).getValue().getBytes(UTF8);
-                inactiveFlowFile = session.write(inactiveFlowFile, new OutputStreamCallback() {
-                    @Override
-                    public void process(final OutputStream out) throws IOException {
-                        out.write(outBytes);
-                    }
-                });
+    private void onTriggerActiveFlow(ProcessContext context, ProcessSession session, boolean wasActive, boolean isClusterScope,
+            long inactivityStartMillis, List<FlowFile> flowFiles) {
+        final ComponentLog logger = getLogger();
+        final boolean shouldReportOnlyOnPrimary = shouldReportOnlyOnPrimary(isClusterScope, context);
+        final boolean shouldThisNodeReport = shouldThisNodeReport(isClusterScope, shouldReportOnlyOnPrimary, context);
 
-                session.getProvenanceReporter().create(inactiveFlowFile);
-                session.transfer(inactiveFlowFile, REL_INACTIVE);
-                logger.info("Transferred {} to 'inactive'", new Object[]{inactiveFlowFile});
-            } else {
-                context.yield();    // no need to dominate CPU checking times; let other processors run for a bit.
+        if (!wasActive) {
+            if (shouldThisNodeReport) {
```

Comment:
`shouldThisNodeReport` local variable could be moved into `if (!wasActive) { ... }`.